### PR TITLE
implement vote against

### PR DESCRIPTION
**Pull request by RuslanProgrammer**

### CODE REVIEW

File: contracts/gov/proposals/DistributionProposal.sol [hunk 1]
```diff
@@ -80,15 +80,19 @@ contract DistributionProposal is IDistributionProposal, Initializable {
...
+            uint256 voterVotesAgainst
+        ) = IGovPool(govAddress).getTotalVotes(proposalId, voter, false);
 
         return
-            totalVoteWeight == 0
+            totalVotesFor == 0
                 ? 0
-                : proposals[proposalId].rewardAmount.ratio(voteWeight, totalVoteWeight);
+                : proposals[proposalId].rewardAmount.ratio(
+                    voterVotesFor - voterVotesAgainst,
```
Comment (Arvolear) [inline contracts/gov/proposals/DistributionProposal.sol:94]: Underflow check

File: contracts/gov/GovPool.sol [hunk 2]
```diff
@@ -382,6 +385,10 @@ contract GovPool is
 
         if (core.settings.earlyCompletion || voteEnd < block.timestamp) {
             if (_quorumReached(core)) {
+                if (!_votesForMoreThanAgainst(core)) {
```
Comment (Arvolear) [inline contracts/gov/GovPool.sol:388]: Check for `against` data

File: contracts/libs/gov-pool/GovPoolExecute.sol [hunk 3]
```diff
@@ -33,32 +33,32 @@ library GovPoolExecute {
...
+        (, , address govValidatorsAddress, ) = GovPool(payable(address(this)))
+            .getHelperContracts();
+        IGovValidators govValidators = IGovValidators(govValidatorsAddress);
 
-        bool validatorsVotingSucceeded = IGovValidators(govValidators).getProposalState(
-            proposalId,
-            false
-        ) == IGovValidators.ProposalState.Succeeded;
+        IGovValidators.ProposalState state = govValidators.getProposalState(proposalId, false);
+        bool validatorsVotingSucceeded = state == IGovValidators.ProposalState.SucceededFor ||
```
Comment (Arvolear) [inline contracts/libs/gov-pool/GovPoolExecute.sol:49]: Useless variable

Comment (RuslanProgrammer) [reply contracts/libs/gov-pool/GovPoolExecute.sol:49]: Check line 68

File: contracts/libs/gov-pool/GovPoolCreate.sol [hunk 4]
```diff
@@ -31,38 +31,33 @@ library GovPoolCreate {
...
+            voteEnd: uint64(block.timestamp + settings.duration),
+            votesFor: 0,
+            votesAgainst: 0,
+            nftPowerSnapshotId: snapshotId
         });
+        proposals[proposalId].descriptionURL = _descriptionURL;
+
+        for (uint256 i; i < actionsOnFor.length; i++) {
+            proposals[proposalId].actionsOnFor.push(actionsOnFor[i]);
+        }
```
Comment (Arvolear) [inline contracts/libs/gov-pool/GovPoolCreate.sol:57]:
```suggestion
        }
        
```

File: test/gov/GovPool.test.js [hunk 5]
```diff
@@ -763,19 +767,24 @@ describe("GovPool", () => {
...
-
-          assert.equal((await getProposalByIndex(1)).descriptionURL, "example.com");
-          assert.equal((await getProposalByIndex(1)).core.votesFor, wei("100"));
-          assert.equal((await getProposalByIndex(2)).core.votesFor, wei("50"));
+          await govPool.vote(1, wei("70"), [], true);
+          await govPool.vote(1, wei("30"), [], false);
+          await govPool.vote(2, wei("50"), [], false);
+
+          let proposal = await getProposalByIndex(1);
+          assert.equal(proposal.descriptionURL, "example.com");
```
Comment (Arvolear) [inline test/gov/GovPool.test.js:775]:
```suggestion
          let proposal = await getProposalByIndex(1);
          
          assert.equal(proposal.descriptionURL, "example.com");
```

File: test/gov/GovPool.test.js [hunk 6]
```diff
@@ -763,19 +767,24 @@ describe("GovPool", () => {
...
+          assert.equal(proposal.core.votesFor, wei("70"));
+          assert.equal(proposal.core.votesAgainst, wei("30"));
+          proposal = await getProposalByIndex(2);
+          assert.equal(proposal.core.votesFor, "0");
```
Comment (Arvolear) [inline test/gov/GovPool.test.js:779]:
```suggestion
          assert.equal(proposal.core.votesAgainst, wei("30"));
          
          proposal = await getProposalByIndex(2);
          
          assert.equal(proposal.core.votesFor, "0");
```

File: test/gov/GovPool.test.js [hunk 7]
```diff
@@ -812,18 +821,23 @@ describe("GovPool", () => {
...
         it("should vote delegated tokens for two proposals", async () => {
-          await govPool.voteDelegated(1, wei("100"), [], true, { from: SECOND });
+          await govPool.voteDelegated(1, wei("70"), [], true, { from: SECOND });
+          await govPool.voteDelegated(1, wei("30"), [], false, { from: SECOND });
           await govPool.voteDelegated(2, wei("50"), [], true, { from: THIRD });
 
-          assert.equal((await getProposalByIndex(1)).core.votesFor, wei("100"));
-          assert.equal((await getProposalByIndex(2)).core.votesFor, wei("50"));
+          let proposal = await getProposalByIndex(1);
+          assert.equal(proposal.core.votesFor, wei("70"));
```
Comment (Arvolear) [inline test/gov/GovPool.test.js:829]:
```suggestion
          let proposal = await getProposalByIndex(1);
          
          assert.equal(proposal.core.votesFor, wei("70"));
```

File: test/gov/GovPool.test.js [hunk 8]
```diff
@@ -812,18 +821,23 @@ describe("GovPool", () => {
...
+          assert.equal(proposal.core.votesAgainst, wei("30"));
+          proposal = await getProposalByIndex(2);
+          assert.equal(proposal.core.votesFor, wei("50"));
```
Comment (Arvolear) [inline test/gov/GovPool.test.js:832]:
```suggestion
          assert.equal(proposal.core.votesAgainst, wei("30"));
          
          proposal = await getProposalByIndex(2);
          
          assert.equal(proposal.core.votesFor, wei("50"));
```

File: test/gov/GovPool.test.js [hunk 9]
```diff
@@ -920,19 +934,33 @@ describe("GovPool", () => {
...
           it("should vote for two proposals", async () => {
             await govPool.vote(1, 0, [1], true);
-            await govPool.vote(2, 0, [2, 3], true);
+            await govPool.vote(2, 0, [2], true);
+            await govPool.vote(2, 0, [3], false);
 
-            assert.equal((await getProposalByIndex(1)).core.votesFor, SINGLE_NFT_COST.toFixed());
-            assert.equal((await getProposalByIndex(2)).core.votesFor, SINGLE_NFT_COST.times(2).plus(1).toFixed());
+            let proposal = await getProposalByIndex(1);
+            assert.equal(proposal.core.votesFor, SINGLE_NFT_COST.toFixed());
```
Comment (Arvolear) [inline test/gov/GovPool.test.js:941]:
```suggestion
            let proposal = await getProposalByIndex(1);
            
            assert.equal(proposal.core.votesFor, SINGLE_NFT_COST.toFixed());
```

File: test/gov/GovPool.test.js [hunk 10]
```diff
@@ -920,19 +934,33 @@ describe("GovPool", () => {
...
+            assert.equal(proposal.core.votesAgainst, "0");
+            proposal = await getProposalByIndex(2);
+            assert.equal(proposal.core.votesFor, SINGLE_NFT_COST.toFixed());
```
Comment (Arvolear) [inline test/gov/GovPool.test.js:944]:
```suggestion
            assert.equal(proposal.core.votesAgainst, "0");
            
            proposal = await getProposalByIndex(2);
            
            assert.equal(proposal.core.votesFor, SINGLE_NFT_COST.toFixed());
```

File: test/gov/GovValidators.test.js [hunk 11]
```diff
@@ -283,19 +284,26 @@ describe("GovValidators", () => {
 
         assert.equal(await validators.addressVoted(1, true, SECOND), wei("40"));
         assert.equal(await validators.addressVoted(1, true, THIRD), wei("0"));
-        assert.equal((await getInternalProposalByIndex(1)).proposal.core.votesFor, wei("40"));
+
+        let core = (await getInternalProposalByIndex(1)).proposal.core;
+        assert.equal(core.votesFor, wei("40"));
```
Comment (Arvolear) [inline test/gov/GovValidators.test.js:289]:
```suggestion
        let core = (await getInternalProposalByIndex(1)).proposal.core;
        
        assert.equal(core.votesFor, wei("40"));
```

File: test/gov/GovValidators.test.js [hunk 12]
```diff
@@ -283,19 +284,26 @@ describe("GovValidators", () => {
...
+        assert.equal(core.votesAgainst, "0");
 
-        await validators.vote(1, wei("60"), true, true, { from: SECOND });
+        await validators.vote(1, wei("60"), true, false, { from: SECOND });
 
         assert.equal(await validators.addressVoted(1, true, SECOND), wei("100"));
         assert.equal(await validators.addressVoted(1, true, THIRD), wei("50"));
-        assert.equal((await getInternalProposalByIndex(1)).proposal.core.votesFor, wei("150"));
+        core = (await getInternalProposalByIndex(1)).proposal.core;
+        assert.equal(core.votesFor, wei("90"));
```
Comment (Arvolear) [inline test/gov/GovValidators.test.js:305]:
```suggestion

        core = (await getInternalProposalByIndex(1)).proposal.core;
        
        assert.equal(core.votesFor, wei("90"));
```

File: test/gov/GovValidators.test.js [hunk 13]
```diff
@@ -305,13 +313,17 @@ describe("GovValidators", () => {
 
         assert.equal(await validators.addressVoted(1, true, SECOND), wei("100"));
         assert.equal(await validators.addressVoted(1, true, THIRD), wei("0"));
-        assert.equal((await getInternalProposalByIndex(1)).proposal.core.votesFor, wei("100"));
+        let core = (await getInternalProposalByIndex(1)).proposal.core;
+        assert.equal(core.votesFor, wei("100"));
```
Comment (Arvolear) [inline test/gov/GovValidators.test.js:317]:
```suggestion
        
        let core = (await getInternalProposalByIndex(1)).proposal.core;
        
        assert.equal(core.votesFor, wei("100"));
```

File: test/gov/GovValidators.test.js [hunk 14]
```diff
@@ -305,13 +313,17 @@ describe("GovValidators", () => {
...
+        assert.equal(core.votesAgainst, "0");
 
-        await validators.vote(1, wei("200"), true, true, { from: THIRD });
+        await validators.vote(1, wei("200"), true, false, { from: THIRD });
 
         assert.equal(await validators.addressVoted(1, true, SECOND), wei("100"));
         assert.equal(await validators.addressVoted(1, true, THIRD), wei("200"));
-        assert.equal((await getInternalProposalByIndex(1)).proposal.core.votesFor, wei("300"));
+        core = (await getInternalProposalByIndex(1)).proposal.core;
+        assert.equal(core.votesFor, wei("100"));
```
Comment (Arvolear) [inline test/gov/GovValidators.test.js:325]:
```suggestion

        core = (await getInternalProposalByIndex(1)).proposal.core;
        
        assert.equal(core.votesFor, wei("100"));
```

File: test/gov/GovValidators.test.js [hunk 15]
```diff
@@ -328,20 +340,24 @@ describe("GovValidators", () => {
...
 
         await validators.vote(1, wei("100"), true, true, { from: SECOND });
-        await validators.vote(1, wei("200"), true, true, { from: THIRD });
+        await validators.vote(1, wei("200"), true, false, { from: THIRD });
 
         assert.equal(await validators.addressVoted(1, true, SECOND), wei("100"));
         assert.equal(await validators.addressVoted(1, true, THIRD), wei("200"));
-        assert.equal((await getInternalProposalByIndex(1)).proposal.core.votesFor, wei("300"));
+        let core = (await getInternalProposalByIndex(1)).proposal.core;
+        assert.equal(core.votesFor, wei("100"));
```
Comment (Arvolear) [inline test/gov/GovValidators.test.js:348]:
```suggestion

        let core = (await getInternalProposalByIndex(1)).proposal.core;
        
        assert.equal(core.votesFor, wei("100"));
```

File: test/gov/GovValidators.test.js [hunk 16]
```diff
@@ -328,20 +340,24 @@ describe("GovValidators", () => {
...
 
-        await validators.vote(3, wei("40"), true, true, { from: SECOND });
+        await validators.vote(3, wei("40"), true, false, { from: SECOND });
         await validators.vote(3, wei("60"), true, true, { from: THIRD });
 
         assert.equal(await validators.addressVoted(3, true, SECOND), wei("40"));
         assert.equal(await validators.addressVoted(3, true, THIRD), wei("60"));
-        assert.equal((await getInternalProposalByIndex(3)).proposal.core.votesFor, wei("100"));
+        core = (await getInternalProposalByIndex(3)).proposal.core;
+        assert.equal(core.votesFor, wei("60"));
```
Comment (Arvolear) [inline test/gov/GovValidators.test.js:359]:
```suggestion

        core = (await getInternalProposalByIndex(3)).proposal.core;
        
        assert.equal(core.votesFor, wei("60"));
```

File: test/gov/GovValidators.test.js [hunk 17]
```diff
@@ -351,19 +367,25 @@ describe("GovValidators", () => {
 
         assert.equal(await validators.addressVoted(2, false, SECOND), wei("40"));
         assert.equal(await validators.addressVoted(2, false, THIRD), wei("0"));
-        assert.equal((await validators.getExternalProposal(2)).core.votesFor, wei("40"));
+        let core = (await validators.getExternalProposal(2)).core;
+        assert.equal(core.votesFor, wei("40"));
```
Comment (Arvolear) [inline test/gov/GovValidators.test.js:371]:
```suggestion

        let core = (await validators.getExternalProposal(2)).core;
        
        assert.equal(core.votesFor, wei("40"));
```

File: test/gov/GovValidators.test.js [hunk 18]
```diff
@@ -351,19 +367,25 @@ describe("GovValidators", () => {
...
+        assert.equal(core.votesAgainst, "0");
 
-        await validators.vote(2, wei("50"), false, true, { from: THIRD });
+        await validators.vote(2, wei("50"), false, false, { from: THIRD });
 
         assert.equal(await validators.addressVoted(2, false, SECOND), wei("40"));
         assert.equal(await validators.addressVoted(2, false, THIRD), wei("50"));
-        assert.equal((await validators.getExternalProposal(2)).core.votesFor, wei("90"));
+        core = (await validators.getExternalProposal(2)).core;
+        assert.equal(core.votesFor, wei("40"));
```
Comment (Arvolear) [inline test/gov/GovValidators.test.js:379]:
```suggestion

        core = (await validators.getExternalProposal(2)).core;
        
        assert.equal(core.votesFor, wei("40"));
```

File: test/gov/GovValidators.test.js [hunk 19]
```diff
@@ -351,19 +367,25 @@ describe("GovValidators", () => {
...
+        assert.equal(core.votesAgainst, wei("50"));
 
         await validators.vote(2, wei("60"), false, true, { from: SECOND });
 
         assert.equal(await validators.addressVoted(2, false, SECOND), wei("100"));
         assert.equal(await validators.addressVoted(2, false, THIRD), wei("50"));
-        assert.equal((await validators.getExternalProposal(2)).core.votesFor, wei("150"));
+        core = (await validators.getExternalProposal(2)).core;
+        assert.equal(core.votesFor, wei("100"));
```
Comment (Arvolear) [inline test/gov/GovValidators.test.js:387]:
```suggestion

        core = (await validators.getExternalProposal(2)).core;
        
        assert.equal(core.votesFor, wei("100"));
```

**Review by Arvolear — APPROVED** [review]
Ok

**Pull request merged**